subprojects/core/src/main/java/org/gradle/internal/progress/DefaultBuildOperationExecutor.java

```diff
@@ -16,133 +16,187 @@
-                ProgressLogger progressLogger;
-                if (operationDetails.getProgressDisplayName() != null) {
-                    progressLogger = progressLoggerFactory.newOperation(DefaultBuildOperationExecutor.class);
-                    progressLogger.setDescription(operationDetails.getDisplayName());
-                    progressLogger.setShortDescription(operationDetails.getProgressDisplayName());
-                    progressLogger.started();
-                } else {
-                    progressLogger = null;
-                }
+                ProgressLogger progressLogger = startProgressLogging(currentOperation);
 
-                LOGGER.debug("Build operation '{}' started", operation.getDisplayName());
+                LOGGER.debug("Build operation '{}' started", description.getDisplayName());
                 try {
-                    result = factory.transform(context);
+                    if (buildOperation instanceof RunnableBuildOperation) {
+                        ((RunnableBuildOperation) buildOperation).run(context);
+                    }
+                    if (buildOperation instanceof CallableBuildOperation) {
+                        result = Cast.uncheckedCast(((CallableBuildOperation) buildOperation).call(context));
+                    }
                 } finally {
                     if (progressLogger != null) {
                         progressLogger.completed();
                     }
                 }
                 if (parent != null && !parent.isRunning()) {
-                    throw new IllegalStateException(String.format("Parent operation (%s) completed before this operation (%s).", parent.operationDetails.getDisplayName(), operationDetails.getDisplayName()));
+                    throw new IllegalStateException(String.format("Parent operation (%s) completed before this operation (%s).",
+                        parent.getDescription().getDisplayName(), description.getDisplayName()));
                 }
             } catch (Throwable t) {
                 context.failed(t);
                 failure = t;
             }
 
             long endTime = timeProvider.getCurrentTime();
-            listener.finished(operation, new OperationResult(startTime, endTime, context.failure, context.result));
+            listener.finished((BuildOperationInternal) description, new OperationResult(currentOperation.getStartTime(), endTime, context.failure, context.result));
 
             if (failure != null) {
                 throw UncheckedException.throwAsUncheckedException(failure, true);
             }
 
-            LOGGER.debug("Build operation '{}' completed", operation.getDisplayName());
+            LOGGER.debug("Build operation '{}' completed", description.getDisplayName());
             return result;
         } finally {
-            if (parent instanceof UnmanagedThreadOperation) {
+            if (parent != null && parent.getDescription().getId() < 0) { //unmanaged thread
```

Comment:
Your are right. I removed it early in the refactoring. But it is good to have, especially for this check. I re-added the type.